### PR TITLE
Migrate 8181 8183 6942 support to rpki rs

### PR DESCRIPTION
**Pull request by timbru**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -50,8 +50,8 @@ rand                  = "^0.8"
 regex                 = { version = "1.5.5", optional = true, default_features = false, features = ["std"] }
 reqwest               = { version = "0.11", features = ["json"] }
 rpassword             = { version = "^5.0", optional = true }
-rpki                  = { version = "0.13.2", features = [ "compat", "repository", "rrdp", "serde" ] }
-# rpki                  = { version = "0.13.2-dev", git = "https://github.com/NLnetLabs/rpki-rs/", branch = "compat-for-prefix-lens", features = [ "compat", "repository", "rrdp", "serde" ] }
+# rpki                  = { version = "0.13.2", features = [ "compat", "repository", "rrdp", "serde" ] }
+rpki                  = { version = "0.14.3-dev", git = "https://github.com/NLnetLabs/rpki-rs/", branch = "support-rfc6492", features = [ "ca", "repository", "rrdp" ] }
```
Comment (ximon18) [inline Cargo.toml:54]: The "ca" feature of "rpki-rs" seems to include the "repository" feature making it redundant here, right?

File: defaults/rules.polar [hunk 2]
```diff
@@ -82,9 +82,9 @@ actor_cannot_access_ca(_: Actor, _: Handle) if false;
 #   2bba. The actor does not have an "inc_cas" attribute (i.e. the actor is not
 #         restricted to certain CAs), _OR_
 #   2bbb. The actor has an "inc_cas" attribute which includes the specified CA
-#         handle (i.e. the CA is included in the set the actor is expicitly
+#         handle (i.e. the CA is included in the set the actor is explicitly
 #         given access to).
-actor_can_access_ca(actor: Actor, ca: Handle) if
+actor_can_access_ca(actor: Actor, ca: PolarHandle) if
```
Comment (ximon18) [inline defaults/rules.polar:87]: Changing this class name is potentially breaking for anyone using their own policy files. Why not either still call it `Handle` (even though it is now a New Type because the underlying `Handle` type has moved out of the crate) or use [`ClassBuilder::name()` ](https://docs.rs/oso/0.12.4/oso/struct.ClassBuilder.html#method.name) to call it `Handle` as far as Oso is concerned, and thus avoid breakage in policy files and to actually refer to it by its real name instead of a "wrapper" name that only exists due to Rusts limitations about implementing traits on external types?

Comment (ximon18) [reply defaults/rules.polar:87]: Also, the references to `Handle` in other `.polar` files have not been updated, e.g. see `abac.polar`.

File: src/cli/options.rs [hunk 3]
```diff
@@ -420,7 +424,7 @@ impl Options {
             Arg::with_name("request")
                 .long("request")
                 .short("r")
-                .help("The location of the RFC8183 Child Request XML file")
+                .help("The location of the idexchange:: Child Request XML file")
```
Comment (ximon18) [inline src/cli/options.rs:427]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 4]
```diff
@@ -447,7 +451,7 @@ impl Options {
     }
 
     fn make_cas_children_response_sc<'a, 'b>(app: App<'a, 'b>) -> App<'a, 'b> {
-        let mut sub = SubCommand::with_name("response").about("Show the RFC8183 Parent Response XML");
+        let mut sub = SubCommand::with_name("response").about("Show the idexchange:: Parent Response XML");
```
Comment (ximon18) [inline src/cli/options.rs:454]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 5]
```diff
@@ -522,7 +526,7 @@ impl Options {
     }
 
     fn make_cas_parents_request_sc<'a, 'b>(app: App<'a, 'b>) -> App<'a, 'b> {
-        let mut sub = SubCommand::with_name("request").about("Show RFC8183 Child Request XML");
+        let mut sub = SubCommand::with_name("request").about("Show idexchange:: Child Request XML");
```
Comment (ximon18) [inline src/cli/options.rs:529]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 6]
```diff
@@ -540,7 +544,7 @@ impl Options {
             Arg::with_name("response")
                 .long("response")
                 .short("r")
-                .help("The location of the RFC8183 Parent Response XML file")
+                .help("The location of the idexchange:: Parent Response XML file")
```
Comment (ximon18) [inline src/cli/options.rs:547]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 7]
```diff
@@ -832,7 +836,7 @@ impl Options {
     }
 
     fn make_cas_repo_request_sc<'a, 'b>(app: App<'a, 'b>) -> App<'a, 'b> {
-        let mut sub = SubCommand::with_name("request").about("Show RFC8183 Publisher Request XML");
+        let mut sub = SubCommand::with_name("request").about("Show idexchange:: Publisher Request XML");
```
Comment (ximon18) [inline src/cli/options.rs:839]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 8]
```diff
@@ -868,7 +872,7 @@ impl Options {
                 .value_name("file")
                 .long("response")
                 .short("r")
-                .help("The location of the RFC8183 Publisher Response XML file")
+                .help("The location of the idexchange:: Publisher Response XML file")
```
Comment (ximon18) [inline src/cli/options.rs:875]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 9]
```diff
@@ -1155,7 +1159,7 @@ impl Options {
                     .value_name("file")
                     .long("request")
                     .short("r")
-                    .help("The location of the RFC8183 Publisher Request XML file")
+                    .help("The location of the idexchange:: Publisher Request XML file")
```
Comment (ximon18) [inline src/cli/options.rs:1162]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 10]
```diff
@@ -1185,7 +1189,7 @@ impl Options {
     }
 
     fn make_publishers_response_sc<'a, 'b>(app: App<'a, 'b>) -> App<'a, 'b> {
-        let mut sub = SubCommand::with_name("response").about("Show RFC8183 Repository Response XML");
+        let mut sub = SubCommand::with_name("response").about("Show idexchange:: Repository Response XML");
```
Comment (ximon18) [inline src/cli/options.rs:1192]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 11]
```diff
@@ -2352,21 +2344,21 @@ pub enum CaCommand {
     Delete(Handle),     // Delete the CA -> let it withdraw and request revocation as well
 
     // Publishing
-    RepoPublisherRequest(Handle), // Get the RFC8183 publisher request
+    RepoPublisherRequest(Handle), // Get the idexchange:: publisher request
```
Comment (ximon18) [inline src/cli/options.rs:2347]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 12]
```diff
@@ -2352,21 +2344,21 @@ pub enum CaCommand {
...
     RepoDetails(Handle),
     RepoUpdate(Handle, RepositoryContact),
     RepoStatus(Handle),
 
     // Parents (to this CA)
-    ChildRequest(Handle), // Get the RFC8183 child request
+    ChildRequest(Handle), // Get the idexchange:: child request
```
Comment (ximon18) [inline src/cli/options.rs:2353]: You probably didn't mean to change this ;-)

File: src/cli/options.rs [hunk 13]
```diff
@@ -2352,21 +2344,21 @@ pub enum CaCommand {
...
     AddParent(Handle, ParentCaReq),
     MyParentCaContact(Handle, ParentHandle),
     ParentStatuses(Handle),
     RemoveParent(Handle, ParentHandle),
     Refresh(Handle), // Refresh with all parents
 
     // Children
-    ParentResponse(Handle, ChildHandle), // Get an RFC8183 parent response for a child
+    ParentResponse(Handle, ChildHandle), // Get an idexchange:: parent response for a child
```
Comment (ximon18) [inline src/cli/options.rs:2361]: You probably didn't mean to change this ;-)

File: src/commons/api/ca.rs [hunk 14]
```diff
@@ -1394,18 +955,25 @@ impl fmt::Display for ParentStatuses {
                         write!(f, "LAST KNOWN Resource Entitlements:")?;
                     }
 
-                    if status.entitlements.is_empty() {
+                    if status.classes.is_empty() {
```
Comment (ximon18) [inline src/commons/api/ca.rs:958]: Are these changes in the set of fields that are output intentional and if so why?

File: src/commons/api/ca.rs [hunk 15]
```diff
@@ -1462,28 +1005,32 @@ impl From<&SigningCert> for ParentStatusIssuingCert {
...
     last_success: Option<Timestamp>,
     all_resources: ResourceSet,
-    entitlements: HashMap<ResourceClassName, KnownEntitlement>,
+
+    // The struct changed - we did not record classes in 0.9.5 and below.
+    // Just default to an empty vec in case this field is missing, and
+    // ignore the 'entitlements' field that used to be there. This will
+    // be updated as soon as the CA synchronizes with its parent again.
+    #[serde(default)]
+    classes: Vec<ResourceClassEntitlements>,
```
Comment (ximon18) [inline src/commons/api/ca.rs:1033]: Why the change from HashMap to Vec? With a Vec you can now presumably get "duplicate" entries in the vec with the same `ResourceClassName` which before would not have been possible.

File: src/commons/api/ca.rs [hunk 16]
```diff
@@ -2815,65 +2191,6 @@ mod test {
...
-        let parent_resources_json = include_str!("../../../test-resources/resources/parent_resources.json");
-        let parent_resources: ResourceSet = serde_json::from_str(parent_resources_json).unwrap();
-
-        let intersection = parent_resources.intersection(&child_resources);
-
-        assert_eq!(intersection, child_resources);
-    }
-
-    #[test]
-    fn resource_set_to_from_string() {
```
Comment (ximon18) [inline src/commons/api/ca.rs:2832]: The other tests appear to have been migrated to `resourceset.rs` in rpki-rs, but I don't see this one?

File: src/commons/api/rrdp.rs [hunk 17]
```diff
@@ -99,13 +98,13 @@ impl PublishElement {
         &self.uri
     }
     pub fn size(&self) -> usize {
-        self.base64.size()
+        self.base64.to_bytes().len()
```
Comment (ximon18) [inline src/commons/api/rrdp.rs:101]: This operation has become much more costly as it now does a base64 decode to get the bytes, does that matter?

File: src/commons/api/rrdp.rs [hunk 18]
```diff
@@ -346,7 +345,8 @@ impl Notification {
                 {
                     // snapshot ref
                     let uri = self.snapshot.uri.to_string();
-                    let a = [("uri", uri.as_str()), ("hash", self.snapshot.hash.as_ref())];
+                    let hash = self.snapshot.hash.to_string();
```
Comment (ximon18) [inline src/commons/api/rrdp.rs:348]: Is this `to_string()` operation costly compared to the prior `.as_ref()` operation? Does it matter?

File: src/commons/api/rrdp.rs [hunk 19]
```diff
@@ -355,10 +355,12 @@ impl Notification {
                     for delta in &self.deltas {
                         let serial = format!("{}", delta.serial);
                         let uri = delta.file_ref.uri.to_string();
+                        let hash = delta.file_ref.hash.to_string();
```
Comment (ximon18) [inline src/commons/api/rrdp.rs:358]: Same question about possible added cost of `to_string()`.

File: src/daemon/ca/manager.rs [hunk 20]
```diff
@@ -1334,48 +1374,111 @@ impl CaManager {
...
+            Ok(bytes) => {
+                cms_logger.reply(&bytes)?;
+                Ok(bytes)
+            }
+        }
+    }
+
+    /// Returns the handle of the local parent for this specific ServiceUri, and the
+    /// configured base (service) uri. Provided that this indeed maps back to this
+    /// same server and it is an rfc6492 style krill uri.
```
Comment (ximon18) [inline src/daemon/ca/manager.rs:1466]: s/uri/URI/, and s/krill/Krill/. Also, elsewhere you have corrected rfcNNNN occurrences to RFC NNNN.

File: src/pubd/repository.rs [hunk 21]
```diff
@@ -1030,30 +1032,37 @@ impl RepositoryAccessProxy {
...
+    pub fn decode_and_validate(
+        &self,
+        publisher: &PublisherHandle,
+        bytes: &[u8],
+    ) -> KrillResult<publication::PublicationCms> {
         let publisher = self.get_publisher(publisher)?;
-        let msg = ProtocolCms::decode(msg, false).map_err(|e| Error::Rfc8181Decode(e.to_string()))?;
-        msg.validate(publisher.id_cert()).map_err(Error::Rfc8181Validation)?;
+        let msg = PublicationCms::decode(bytes).map_err(Error::Rfc8181)?;
+        msg.validate(publisher.id_cert()).map_err(Error::Rfc8181)?;
```
Comment (ximon18) [inline src/pubd/repository.rs:1047]: Why was this changed from `Error::Rfc8181Validation` to `Error::Rfc8181` given that the context is validation?

File: src/test.rs [hunk 22]
```diff
@@ -270,11 +276,11 @@ pub async fn parent_contact(handle: &Handle, child: &ChildHandle) -> ParentCaCon
     .await
     {
         ApiResponse::ParentCaContact(contact) => contact,
-        _ => panic!("Expected RFC8183 parent response"),
+        _ => panic!("Expected idexchange:: parent response"),
```
Comment (ximon18) [inline src/test.rs:279]: This probably wasn't what you intended, right?

File: src/upgrades/mod.rs [hunk 23]
```diff
@@ -463,7 +453,7 @@ fn record_preexisting_openssl_keys_in_signer_mapper(config: Arc<Config>) -> Resu
 
             for entry in dir_iter {
                 let entry = entry.map_err(|err| {
-                    UpgradeError::IoError(KrillIoError::new(
+                    PrepareUpgradeError::IoError(KrillIoError::new(
```
Comment (ximon18) [inline src/upgrades/mod.rs:456]: Nice catch.

**Review by ximon18 — APPROVED** [review]
LGTM. I haven't tried to compile the code or run the tests, and have only checked some of the code now depended on from the rpki-rs crate.

File: Cargo.toml (hunk 1 above)
Comment (timbru) [reply Cargo.toml]: done

File: src/cli/options.rs (hunk 5 above)
Comment (timbru) [reply src/cli/options.rs]: nope, fixed

File: src/cli/options.rs (hunk 6 above)
Comment (timbru) [reply src/cli/options.rs]: nope, fixed

File: src/cli/options.rs (hunk 10 above)
Comment (timbru) [reply src/cli/options.rs]: nope, fixed

File: src/cli/options.rs (hunk 12 above)
Comment (timbru) [reply src/cli/options.rs]: nope, fixed

File: src/cli/report.rs [hunk 24]
```diff
@@ -3,6 +3,8 @@ use std::str::FromStr;
 
 use serde::Serialize;
 
+use rpki::ca::idexchange;
```
[the remark this answers is not in the document]
Comment (timbru) [reply src/cli/report.rs:6]: I would rather take the pain of fixing it now. And it's already done.

File: src/commons/api/ca.rs (hunk 14 above)
Comment (timbru) [reply src/commons/api/ca.rs:957]: Main difference is that instead of interpreting some bits we now just keep the complete response as the parent sent it. I think this will be better for debugging purposes.

It turned out that doing this was easier than changing the existing functions to use the new rpki-rs code.

In any case this is pretty much hidden, debug info.

File: src/commons/api/ca.rs (hunk 15 above)
Comment (timbru) [reply src/commons/api/ca.rs:1032]: The resource class name is part of `ResourceClassEntitlements`. It's true that duplicates could occur, but only if the parent sent us duplicates - which would be an error.

File: src/cli/options.rs (hunk 11 above)
Comment (timbru) [reply src/cli/options.rs]: fixed

File: src/test.rs (hunk 22 above)
Comment (timbru) [reply src/test.rs]: fixed

File: tests/testbed.rs [hunk 25]
```diff
@@ -90,7 +94,8 @@ async fn add_and_remove_certificate_authority() {
     )
     .await
     .unwrap();
-    assert!(parent_response_xml.starts_with("<parent_response "));
+
+    assert!(parent_response_xml.starts_with("\n<parent_response "));
```
[the remark this answers is not in the document]
Comment (timbru) [reply tests/testbed.rs:101]: the new lib (in rpki-rs) starts with a newline before the first element, this is not important, but this test insisted..

File: defaults/rules.polar (hunk 2 above)
Comment (timbru) [reply defaults/rules.polar]: Done! Good point.

File: src/commons/api/rrdp.rs (hunk 17 above)
Comment (timbru) [reply src/commons/api/rrdp.rs]: Now using better size_approx from rpki-rs

File: src/commons/api/rrdp.rs [hunk 26]
```diff
@@ -130,28 +129,28 @@ impl From<publication::Publish> for PublishElement {
...
-    pub fn hash(&self) -> &HexEncodedHash {
+    pub fn hash(&self) -> &Hash {
         &self.hash
     }
     pub fn base64(&self) -> &Base64 {
         &self.base64
     }
     pub fn size(&self) -> usize {
-        self.base64.size()
+        self.base64.to_bytes().len()
```
[the remark this answers is not in the document]
Comment (timbru) [reply src/commons/api/rrdp.rs]: Now using better size_approx from rpki-rs
